GitUI/CommandsDialogs/BrowseDialog/GitStatusMonitor.cs

```diff
@@ -384,86 +311,99 @@ private void Update()
                 UICommandsSource.UICommands.RepoChangedNotifier.IsLocked ||
                 (GitVersion.Current.RaceConditionWhenGitStatusIsUpdatingIndex && Module.IsRunningGitProcess()))
             {
-                _statusIsUpToDate = false; // tell that computed status isn't up to date
+                _statusIsUpToDate = false; // schedule new update when command is finished
                 return;
             }
 
+            _workTreeWatcher.EnableRaisingEvents = true;
             _commandIsRunning = true;
             _statusIsUpToDate = true;
             _previousUpdateTime = Environment.TickCount;
 
+            // Schedule update every 5 min, even if we don't know that anything changed
+            CalculateNextUpdateTime(PeriodicUpdateInterval);
+
+            // capture a consistent state in the main thread
+            var statusCancellationToken = _statusCancellationToken;
+            IGitModule module = Module;
+
             ThreadHelper.JoinableTaskFactory.RunAsync(
                 async () =>
                 {
                     try
                     {
-                        await TaskScheduler.Default;
-
-                        _ignoredFilesPending = _ignoredFilesAreStale;
-
-                        // git-status with ignored files when needed only
-                        var cmd = GitCommandHelpers.GetAllChangedFilesCmd(!_ignoredFilesPending, UntrackedFilesMode.Default, noLocks: true);
-                        var output = Module.RunGitCmd(cmd);
-                        var changedFiles = GitCommandHelpers.GetStatusChangedFilesFromString(Module, output);
+                        try
+                        {
+                            await TaskScheduler.Default;
+                            if (!statusCancellationToken.IsCancellationRequested)
+                            {
+                                var cmd = GitCommandHelpers.GetAllChangedFilesCmd(true, UntrackedFilesMode.Default, noLocks: true);
+                                var output = module.RunGitCmd(cmd);
+                                var changedFiles = GitCommandHelpers.GetStatusChangedFilesFromString(module, output);
 
-                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(statusCancellationToken);
 
-                        UpdatedStatusReceived(changedFiles);
+                                UpdatedStatusReceived(changedFiles);
+                            }
+                        }
+                        catch
+                        {
+                            try
+                            {
+                                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(statusCancellationToken);
+                            }
+                            finally
+                            {
+                                CurrentStatus = GitStatusMonitorState.Stopped;
```

Comment:
Performing this in the `finally` block does not guarantee that we are on the main thread.

Reply:
Intentionally done like that even if not threadsafe. The worst that could happen is that the status is overwritten, this writing should be atomic.
Changed the handling to not introduce a bad pattern.